api_app/services/cost_service.py

```diff
@@ -45,29 +46,106 @@ class SubscriptionNotSupported(Exception):
     """Raised when subscription does not support cost management"""
 
 
+class TooManyRequests(Exception):
+    """Raised when cost management api is being throttled, retry after given number of seconds"""
+    retry_after: int
+
+    def __init__(self, retry_after: int, *args: object) -> None:
+        super().__init__(*args)
+        self.retry_after = retry_after
+
+
+class ServiceUnavailable(Exception):
+    """Raised when cost management is unavaiable, retry after given number of seconds"""
+    retry_after: int
+
+    def __init__(self, retry_after: int, *args: object) -> None:
+        super().__init__(*args)
+        self.retry_after = retry_after
+
+
+class CostCacheItem():
+    """Holds cost qery result and time to leave for storing in cache"""
+    result: QueryResult
+    ttl: datetime
+
+    def __init__(self, item: QueryResult, ttl: datetime) -> None:
+        self.result = item
+        self.ttl = ttl
+
+
+# make sure CostService is singleton
+@lru_cache(maxsize=None)
```

Comment:
I've generally seen lru_cache used with a class factory (i.e. a method that returns an instance) rather than on a class definition. 

In fact, maxsize=None generally means it has an ever-growing cache.

Do you have something I can read that explains how this makes it a singleton?

Reply:
Ah...ok...so does this work because the __init__ has no parameters?

Reply:
exactly, it works because there are no parameters.
at the beginning I've instantiated it on the on_start event of the app and stored it inside the app.state,
but then I saw this implementation of lru_cache and I've changed the code,
I'll search for class factory example, probably that are nicer ways to implement that in python with less potential bugs.

Reply:
A simple example showing both approaches might be...

```python
from functools import lru_cache
import datetime

@lru_cache(maxsize=None)
class Test:

    the_time = None
    
    def __init__(self) -> None:
        self.the_time = datetime.datetime.now()
    
    def GetCounter(self) -> int:
        return self.the_time

@lru_cache(maxsize=None)
def test_factory() -> Test:
    return Test() 
```

With the factory approach it's a little more obvious how the caching is working.

Reply:
I've added a factory method, and it works, can you please take a look?

Reply:
Looks good. I suspect you can then remove the LRU from the class definition, but it's fine to leave it there as well I guess.